src/main/scala/shine/OpenCL/compilation/HostManagedBuffers.scala

```diff
@@ -0,0 +1,271 @@
+  ): (Phrase[CommType], Metadata) = {
+    val meta = Metadata(mutable.Set(), mutable.Set(), mutable.Set(), mutable.Set())
+    val p2 = VisitAndRebuild(p, Visitor(allocs, managed, meta))
+    (p2, meta)
+  }
+
+  private case class Visitor(allocs: Set[Identifier[_ <: PhraseType]],
+                             managed: mutable.Map[Identifier[_ <: PhraseType], AccessFlags],
+                             metadata: Metadata) extends VisitAndRebuild.Visitor {
+    override def phrase[T <: PhraseType](p: Phrase[T]): Result[Phrase[T]] =
+      p match {
+        case dpia.Assign(_, lhs, rhs) =>
+          collectWrites(lhs, metadata.host_writes)
+          collectReads(rhs, allocs, metadata.host_reads)
+          Stop(p)
+        case ocl.KernelCallCmd(_, _, _, out, in) =>
+          in.foreach(collectReads(_, allocs, metadata.device_reads))
+          collectWrites(out, metadata.device_writes)
+          ((out, DEVICE_WRITE) +: in.map(_ -> DEVICE_READ)).foreach {
+            case (i: Identifier[_], a) => recordManagedAccess(managed, i, a)
+            case (Proj1(i: Identifier[_]), a) => recordManagedAccess(managed, i, a)
+            case (Proj2(i: Identifier[_]), a) => recordManagedAccess(managed, i, a)
+            case (Natural(_), _) =>
+            case (unexpected, _) => throw new Exception(s"did not expect $unexpected")
+          }
+          Stop(p)
+        case dpia.Seq(a, b) =>
+          val (a2, am) = analyzeAndInsertHostExecution(a, allocs, managed)
+          val (b2, bm) = insertHostExecutions(am, allocs, managed, b)
+          metadata.host_writes ++= bm.host_writes
+          metadata.host_reads ++= bm.host_reads
+          metadata.device_writes ++= bm.device_writes
+          metadata.device_reads ++= bm.device_reads
+          Stop(dpia.Seq(a2, b2))
+        case _ => Continue(p, this)
+      }
+  }
+
+  def optionallyManaged(i: Identifier[_ <: PhraseType])
+  : Option[(Identifier[_ <: PhraseType], DataType)]
+  = i.`type` match {
+    case ExpType(_: ScalarType, _) | ExpType(_: IndexType, _) => None
+    case ExpType(dt, a) => Some(Identifier("m" + i.name, ExpType(ManagedBufferType(dt), a)), dt)
+    case AccType(_: ScalarType) => None
+    case AccType(dt) => Some(Identifier("m" + i.name, AccType(ManagedBufferType(dt))), dt)
+    case PhrasePairType(ExpType(_: ScalarType, _), AccType(_)) => None
+    case PhrasePairType(ExpType(dt, a), AccType(_)) => Some(Identifier("m" + i.name,
+      PhrasePairType(ExpType(ManagedBufferType(dt), a), AccType(ManagedBufferType(dt)))), dt)
+    case _ => throw new Exception("this should not happen")
+  }
+
+  private def insertManagedBuffers(
+    managed: mutable.Map[Identifier[_ <: PhraseType], AccessFlags]
+  ): Phrase[CommType] => Phrase[CommType] = p => {
+    val managed2 = managed.iterator.flatMap { case (i, a) =>
+      optionallyManaged(i).map(om => i -> (a, om._1))
+    }.toMap
+    VisitAndRebuild(p, Visitor2(managed2))
+  }
+
+  private case class Visitor2(
```

Comment:
Can we try to find more expressive names for this (and the Visitor above)?

Reply:
`InsertHostExecutionsVisitor`/`InsertManagedBuffersVisitor` modulo terminology changes we might do.

Reply:
:+1: